src/R/Components/Test/Plots/PlotIntegrationTest.cs

```diff
@@ -603,25 +1057,56 @@ public class PlotIntegrationTest : IAsyncLifetime {
             }
         }
 
+        private async Task WaitForDeviceCreatedAsync(Func<Task> action) {
+            var eas1 = new EventTaskSource<IRPlotManager>((o, h) => o.DeviceCreateMessageReceived += h, (o, h) => o.DeviceCreateMessageReceived -= h);
+            var eas2 = new EventTaskSource<IRPlotManager>((o, h) => o.ActiveDeviceChanged += h, (o, h) => o.ActiveDeviceChanged -= h);
+            var task1 = eas1.Create(_workflow.Plots);
+            var task2 = eas2.Create(_workflow.Plots);
+            await action();
+            await task1;
+            await task2;
+        }
+
+        private async Task WaitForDeviceChangedAsync(Func<Task> action) {
+            var eas = new EventTaskSource<IRPlotManager>((o, h) => o.ActiveDeviceChanged += h, (o, h) => o.ActiveDeviceChanged -= h);
+            var task = eas.Create(_workflow.Plots);
+            await action();
+            await task;
+        }
+
+        private async Task WaitForPlotAsync(IRPlotDeviceViewModel vm, Func<Task> action) {
+            var eas = new EventTaskSource<IRPlotDeviceViewModel>((o, h) => o.PlotChanged += h, (o, h) => o.PlotChanged -= h);
+            var task = eas.Create(vm);
+            await action();
+            await task;
+        }
+
         private async Task WaitForPlotAsync(Func<Task> action) {
-            var eas = new EventTaskSource<IRPlotManager>((o, h) => o.PlotChanged += h, (o, h) => o.PlotChanged -= h);
-            var plotChangedTask = eas.Create(_workflow.Plots);
+            var eas = new EventTaskSource<IRPlotManager>((o, h) => o.PlotMessageReceived += h, (o, h) => o.PlotMessageReceived -= h);
+            var task = eas.Create(_workflow.Plots);
+            await action();
+            await task;
+        }
+
+        private async Task WaitForMutated(Func<Task> action) {
+            var eas = new EventTaskSource<IRSession, EventArgs>((o, h) => o.Mutated += h, (o, h) => o.Mutated -= h);
```

Comment:
With that, you don't have to add these helper methods, which should simplify test code as well

---
In reply to: [76331453](https://github.com/Microsoft/RTVS/pull/2199#discussion_r76331453) [](ancestors = 76331453)